haystack/errors.py

```diff
@@ -1,8 +1,84 @@
 # coding: utf8
-"""Custom Errors for Haystack stacks"""
+"""Custom Errors for Haystack"""
 
+from typing import Optional
+from jsonschema.exceptions import ValidationError
 
-class DuplicateDocumentError(ValueError):
+
+class HaystackError(Exception):
+    """
+    Any error generated by Haystack.
+
+    This error wraps its source transparently in such a way that its attributes
+    can be accessed directly: if the original error has a `message` attribute,
+    `HaystackError.message` will exist and have the expected content.
+
+    Give the source error as the `source` parameter to enable this behavior.
+    Subclasses of HaystackError might enforce specific error classes as their source.
+    """
+
+    def __init__(self, message: str = "", source: Optional[Exception] = None, docs_link: Optional[str] = None):
```

Comment:
I'll test this though. I introduced this complex error printing system because the YAML validator likes to spit out the entire schema when something is wrong top-level, and I want to hide it. Using `from` won't help, as it always prints both stacktraces.

I need to think about this... maybe we can fix it in a separate PR though? Let's open an issue if so.

Reply:
Ah ok, makes sense. I'd rather introduce a special Error for validation that cuts the output instead of making all Errors wrap the original_exc manually.